persistable state docs (#105202)
elastic/kibana: dev_docs/key_concepts/persistable_state.mdx

---
id: kibDevDocsPersistableStateIntro
slug: /kibana-dev-docs/persistable-state-intro
title: Persistable State
summary: Persitable state is a key concept to understand when building a Kibana plugin.
date: 2021-02-02
tags: ['kibana','dev', 'contributor', 'api docs']
---

 “Persistable state” is developer-defined state that supports being persisted by a plugin other than the one defining it. Persistable State needs to be serializable and the owner can/should provide utilities to migrate it, extract and inject references from/to it, as well as telemetry collection utilities.
 
## Exposing state that can be persisted

> Any plugin that exposes state or configuration that another plugin might persist should implement `PersistableStateService`
> interface on their `setup` contract. This will allow plugins persisting the state to easily access migrations and other utilities.
> 
> example: Data plugin allows you to generate filters. Those filters can be persisted by applications in their saved 
> objects or in the URL. In order to allow apps migrating the filters in case the structure changes in the future Data 
> plugin implements `PersistableStateService` on `data.query.filterManager`.
> 

Reviewer: So we can get this merged, I think we should highlight some of the risks with the current architecture. Maybe...

```suggestion
Any plugin that exposes state that another plugin can persist should implement `PersistableStateService`
interface on their `setup` contract. This will allow plugins persisting the state to access migrations and other utilities. 

There is currently no obvious way for a plugin to know which state is safe to persist. The developer must manually look for a matching `PersistableStateService`, or ad-hoc provided migration utilities (as is the case with Rule Type Parameters).

We already have situations where state is being persisted, even though no migration utilities are exposed. For example, filters from the data plugin are persisted in URLs and saved objects, but don't currently expose a `PersistableStateService`. This is something we will need to fix.

In the future, we hope to make it easier for producers of state to understand when they need to write a migration with changes, and also make it easier for consumers of such state, to understand whether it is safe to persist. 
```

## exposing state that can be persisted but is not owned by plugin exposing it (registry)

Any plugin that owns collection of items whose state/configuration can be persisted should implement `PersistableStateService`
interface on their `setup` contract as well as make each item in the collection implement `PersistableStateDefinition` interface.

> example: Embeddable plugin owns the registry of embeddable factories to which other plugins can register new embeddable factories. Dashboard plugin

Reviewer: ```suggestion
Example: Embeddable plugin owns the registry of embeddable factories to which other plugins can register new embeddable factories. Dashboard plugin
```

> interface on their `setup` contract as well as make each item in the collection implement `PersistableStateDefinition` interface.

Reviewer 2: > make each item in the collection implement `PersistableStateDefinition` interface.

Is it the plugin that is registering the item that will have to implement the `PersistableStateDefinition` interface on the item? Or is it the plugin that owns the collection?

Author: its the plugin providing the registry that defines how the items that are added to the registry look like

stores a bunch of embeddable panels input in its saved object and URL. Embeddable plugin setup contract implements `PersistableStateService`
interface and each `EmbeddableFactory`  needs to implement `PersistableStateDefinition` interface.

note: if your plugin doesn't expose the state and is thus the only one storing state but still owning the registry
of items whose state it doesn't own then the items should still implement `PersistableStateDefinition` but registry
plugin doesn't need to implement `PersistableStateService` interface.

> ## storing persistable state as part of saved object

Reviewer: ```suggestion
## Storing persistable state as part of saved object
```

> note: if your plugin doesn't expose the state and is thus the only one storing state but still owning the registry

Reviewer 2: Might be worth rewording this to something like:

`if your plugin doesn't expose state, then your plugin doesn't need to implement `PersistableStateService` interface on the `setup` contract. If your plugin provide registration functionality for items it will not own the state for, the items should still implement `PersistableStateDefinition`.

Any plugin that stores any persistable state as part of their saved object should make sure that its saved object migration 
and reference extraction and injection methods correctly use the correct `PersistableStateService` implementation.

> For migrations saved object definition `migration` key should be a function, which returns a list of migrations. `getAllMirations`
> function of `PersistableStateService` should be called in there to get an object with migrations which should be merged with
> any migrations provided by the plugin.

Reviewer: ```suggestion
If a saved object contains state from another plugin, the <DocLink id="kibCoreSavedObjectsPluginApi" section="iaf33fb81-e64a-11eb-8631-9d6e17a74ca4" text-"SavedObjectType.migrations"/> parameter should be a function, which returns a list of migrations. 
 These migration functions will need to merge the migration functions returned from <DocLInk id="" section="" text="PersistableStateServices.getAllMigrations"/> with any migrations provided by the plugin.
```

^^ I didn't finish the DocLinks, or verify they are correct, but hopefully you get the idea and can fill them in.

For reference injection and extraction matching methods on PersistableStateService should be called and merged with plugins own array of references.

> ## available state operations

Reviewer: ```suggestion
## Available state operations
```

> ### storing persistable state as part of URL

Reviewer: ```suggestion
### Storing persistable state as part of URL
```

> When storing persistable stae as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method

Reviewer: ```suggestion
When storing persistable state as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method
```

>  These migration functions will need to merge the migration functions returned from with any migrations provided by the plugin.

Reviewer 2: This sentence is a bit confusing to me. Is the `These migration functions` referring to the function parameter in the previous sentence or the `list of migrations` that the previous sentence is talking about?

Is this trying to say: the parameter function needs to call a call the plugin that owns the state it is trying to persistent and merge the result of that call into an array of migration functions with any migration logic that the parameter function will provide for state it also exposes?

An example might be helpful here.

> For reference injection and extraction matching methods on PersistableStateService should be called and merged with plugins own array of references.

Reviewer 2: A code example here might be helpful.

> When storing persistable state as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method

Reviewer 2: > When storing persistable state as part of URL you must make sure your URL is versioned

Do we have suggestions on how to version the URL? Is it simply adding some field that has a version value in it?

Author: nothing, its completely up to the app how it does that.

> of `PersistableStateService` should be called, which will given the original version of state migrate it to latest.

Reviewer 2: > of `PersistableStateService` should be called, which will given the original version of state migrate it to latest.

Maybe switch this to:

of `PersistableStateService` should be called, which will migrate the state from its original version to the latest.

 
### Extraction/Injection of References

In order to support import and export, and space-sharing capabilities, Saved Objects need to explicitly list any references they contain to other Saved Objects.

> To support persisting your state in saved objects owned by another plugin, the `extract` and `inject` methods of Persistable State interface should be implemented.

Reviewer 2: Including a link to the interface and an example implementation of these methods might be helpful.

Reviewer: ```suggestion
To support persisting your state in saved objects owned by another plugin, the <DocLink id="kibKibanaUtilsPluginApi" section="def-common.PersistableState.extract" text="`extract`"/> and <DocLink id="kibKibanaUtilsPluginApi" section="def-common.PersistableState.inject" text="`inject`"/> methods of Persistable State interface should be implemented.
```

<DocLink id="kibDevTutorialSavedObject" section="references" text="Learn how to define Saved Object references"/>

### Migrations and Backward compatibility

> As your plugin evolves, you may need to change your state in a breaking way. If that happens, you should write a migration to upgrade the any state that existed prior to the change.

Reviewer 2: nit: `the any` -> `the` or `any`

migration to upgrade the any state that existed prior to the change. -> migration to upgrade **_the_** state that existed prior to the change.

<DocLink id="kibDevTutorialSavedObject" section="migrations" text="How to write a migration"/>.

## Telemetry

> You might want to collect state about how your state is used. If that is the case you should implement the telemetry method of Persistable State interface.

Reviewer 2: An example here might be helpful.

`You might want to collect state about how your state is used.` -> `You might want to collect statistics about how your state is used.`

Author: we don't have an example for this yet, will be added in the future.

Reviewer: ```suggestion
You might want to collect statistics about how your state is used. If that is the case you should implement the telemetry method of Persistable State interface. 
```

stores a bunch of embeddable panels input in its saved object and URL. Embeddable plugin setup contract implements `PersistableStateService`
interface and each `EmbeddableFactory`  needs to implement `PersistableStateDefinition` interface.

note: if your plugin doesn't expose the state and is thus the only one storing state but still owning the registry
of items whose state it doesn't own then the items should still implement `PersistableStateDefinition` but registry
plugin doesn't need to implement `PersistableStateService` interface.

## Storing persistable state as part of saved object

Any plugin that stores any persistable state as part of their saved object should make sure that its saved object migration 
and reference extraction and injection methods correctly use the correct `PersistableStateService` implementation.

If a saved object contains state from another plugin, the <DocLink id="kibCoreSavedObjectsPluginApi" section="iaf33fb81-e64a-11eb-8631-9d6e17a74ca4" text-"SavedObjectType.migrations"/> parameter should be a function, which returns a list of migrations. 

>  These migration functions will need to merge the migration functions returned from <DocLInk id="" section="" text="PersistableStateServices.getAllMigrations"/> with any migrations provided by the plugin.

Author: those docs don't exist yet so i'll remove it

>  These migration functions will need to merge the migration functions returned from <DocLInk id="" section="" text="PersistableStateServices.getAllMigrations"/> with any migrations provided by the plugin.

Reviewer: I'm talking about linking to the API docs for that function. Do the API docs need to be re-generated? If so, I can do that. I'm assuming the function exists, otherwise, we shouldn't be adding documentation recommending its usage. :)

For reference injection and extraction matching methods on PersistableStateService should be called and merged with plugins own array of references.

## Available state operations

### Storing persistable state as part of URL

When storing persistable state as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method
of `PersistableStateService` should be called, which will given the original version of state migrate it to latest.

 
### Extraction/Injection of References

In order to support import and export, and space-sharing capabilities, Saved Objects need to explicitly list any references they contain to other Saved Objects.

> To support persisting your state in saved objects owned by other plugin `extract` and `inject` method of Persistable State interface should be implemented.

Reviewer: ```suggestion
To support persisting your state in saved objects owned by another plugin, the `extract` and `inject` methods of Persistable State interface should be implemented.
```

---
id: kibDevDocsPersistableStateIntro
slug: /kibana-dev-docs/persistable-state-intro
title: Persistable State
summary: Persitable state is a key concept to understand when building a Kibana plugin.
date: 2021-02-02
tags: ['kibana','dev', 'contributor', 'api docs']
---

"Persistable State" is developer defined part of state or configuration that can be(is ment to be) persisted by other
 plugin that the one providing it. Persistable State needs to be serializable and owner can/should provide utilities
 to migrate it, extract and inject references from/to it as well as collect telemetry.
 
## exposing state that can be persisted

Any plugin that exposes state or configuration that another plugin might persist should implement `PersistableStateService`

> interface on their `setup` contract. This will allow plugins persisting the state to easily access migrations and other utilities.

Author: 1. if lens allows using lens without embeddable, then it should. To me it would make more sense that lens would only be exposed thru embeddable.

2. consumer: i think we should mention it in the docs, for example documentation for embeddbles should make it clear that embeddable input is persistable state. 

3. we were discussing joining migrate+inject into deserialize and extract into serialize before, but i think one of the problems was that this will not always be the same: on the server we'll need to make extraction and migration separate (due to saved object part) and on the client we won't be extracting references at all (for urls). There is also this thing about naming, the methods would not serialize/deserialize so most likely they should not be called so to avoid any confusion

> interface on their `setup` contract. This will allow plugins persisting the state to easily access migrations and other utilities.

Reviewer: > 1.

Checked the code and it looks like Lens doesn't expose access to it directly from its start contract, so stgm. 

> 2. 

I think it's okay to say that producers should leverage comments to make the connection between persisted state and persistable state service clear, but I don't think that's a good long term solution.

> 3.

For server, in that case, we could still expose `getMigrationFunctions` and `inject` separately. Could we expose `deserialize` and `serialize` on both client and server though as a convenience function? Or perhaps just on client? Having `serialize` as part of the PersistableStateService means we can rely on typescript, not code comments, to help consumers understand the connection.

---
id: kibDevDocsPersistableStateIntro
slug: /kibana-dev-docs/persistable-state-intro
title: Persistable State
summary: Persitable state is a key concept to understand when building a Kibana plugin.
date: 2021-02-02
tags: ['kibana','dev', 'contributor', 'api docs']
---

>  “Persistable state” is developer-defined state that supports being persisted by a plugin other than the one defining it. Persistable State needs to be serializable and the owner can/should provide utilities to migrate it, extract and inject references from/to it, as well as telemetry collection utilities.
>  

Reviewer: ```suggestion
 “Persistable state” is developer-defined state that supports being persisted by a plugin other than the one defining it. Persistable State needs to be serializable and the owner can/should provide utilities to migrate it, extract and inject any <DocLink id="kibDevDocsSavedObjectsIntro" section="references" text="references to Saved Objects"/> it may contain, as well as telemetry collection utilities.
 
```

Occurred to me, a reader may not know what these references are.

## Exposing state that can be persisted

> Any plugin that exposes state that another plugin might persist should implement `PersistableStateService` interface on their `setup` contract. This will allow plugins persisting the state to easily access migrations and other utilities.

Reviewer: ```suggestion
Any plugin that exposes state that another plugin might persist should implement <DocLink id="kibKibanaUtilsPluginApi " section="def-common.PersistableStateService" text="`PersistableStateService`"/> interface on their `setup` contract. This will allow plugins persisting the state to easily access migrations and other utilities.
```

Example: Data plugin allows you to generate filters. Those filters can be persisted by applications in their saved 

> objects or in the URL. In order to allow apps to migrate the filters in case the structure changes in the future, the Data plugin implements `PersistableStateService` on `data.query.filterManager`.
> 

Reviewer: ```suggestion
objects or in the URL. In order to allow apps to migrate the filters in case the structure changes in the future, the Data plugin implements `PersistableStateService` on <DocLink id="kibDataQueryPluginApi " section="def-public.FilterManager" text="`data.query.filterManager`"/>.

```

note: There is currently no obvious way for a plugin to know which state is safe to persist. The developer must manually look for a matching `PersistableStateService`, or ad-hoc provided migration utilities (as is the case with Rule Type Parameters).
In the future, we hope to make it easier for producers of state to understand when they need to write a migration with changes, and also make it easier for consumers of such state, to understand whether it is safe to persist.

## Exposing state that can be persisted but is not owned by plugin exposing it (registry)

Any plugin that owns collection of items (registry) whose state/configuration can be persisted should implement `PersistableStateService`

> interface on their `setup` contract and each item in the collection should implement `PersistableStateDefinition` interface.
> 

Reviewer: ```suggestion
interface on their `setup` contract and each item in the collection should implement <DocLink id="kibKibanaUtilsPluginApi" section="def-common.PersistableStateDefinition" text="`PersistableStateDefinition`"/> interface.

```

Example: Embeddable plugin owns the registry of embeddable factories to which other plugins can register new embeddable factories. Dashboard plugin
stores a bunch of embeddable panels input in its saved object and URL. Embeddable plugin setup contract implements `PersistableStateService`
interface and each `EmbeddableFactory`  needs to implement `PersistableStateDefinition` interface.

Embeddable plugin exposes this interfaces:
```
// EmbeddableInput implements Serializable

export interface EmbeddableRegistryDefinition extends PersistableStateDefinition<EmbeddableInput> {
  id: string;
  ...
}

export interface EmbeddableSetup extends PersistableStateService<EmbeddableInput>;
```

> note: if your plugin doesn't expose the state (is the only one storing state) plugin doesn't need to implement `PersistableStateService` interface.

Reviewer: ```suggestion
Note: if your plugin doesn't expose the state (it is the only one storing state), the plugin doesn't need to implement the `PersistableStateService` interface.
```

If the state your plugin is storing can be provided by other plugins (your plugin owns a registry) items in that registry still need to implement `PersistableStateDefinition` interface.

## Storing persistable state as part of saved object

Any plugin that stores any persistable state as part of their saved object should make sure that its saved object migration 
and reference extraction and injection methods correctly use the matching `PersistableStateService` implementation for the state they are storing.

Take a look at [example saved object](https://github.com/elastic/kibana/blob/master/examples/embeddable_examples/server/searchable_list_saved_object.ts#L32) which stores an embeddable state. Note how the `migrations`, `extractReferences` and `injectReferences` are defined.

## Storing persistable state as part of URL

When storing persistable state as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method
of `PersistableStateService` should be called, which will migrate the state from its original version to latest.

> note: Currently there is no recommended way on how to store version in url and its up to every application to decide on how to implement that.

Author: we leave it completely up to the plugins how they want to do this.
